### PR TITLE
Generic logging framework for SentenceTransformer and CrossEncoder

### PR DESCRIPTION
**Pull request by zachschillaci27**
Following the contributions of [skewwhiff](https://github.com/skewwhiff) and [aaronrmm](https://github.com/aaronrmm), I'm preparing a new PR to supersede #1606. This is a great logging feature to include - so please let me know if anything else is needed to merge!

### CODE REVIEW

**Comment by EricSchuMa** [comment]
This is very useful. Would love for this to be merged soon.

**Comment by lzjever** [comment]
I've been using your patch for months, it works well.

**Comment by shreyasavant** [comment]
@zachschillaci27 How can I use your patch? Apologies if this is a primitive question, I have tried all the PRs in this topic and have been struggling to integrate it with a TSDAE implementation.

**Comment by zachschillaci27** [comment]
> @zachschillaci27 How can I use your patch? Apologies if this is a primitive question, I have tried all the PRs in this topic and have been struggling to integrate it with a TSDAE implementation.

Not a problem. Basically you will need to build `sentence-transformers` off of my branch.

First off, clone the repository
```
git clone  git@github.com:zachschillaci27/sentence-transformers.git
```
replace with https://github.com/zachschillaci27/sentence-transformers.git if you use https authentication.

Then, `cd` into the repository and checkout my branch
```
cd sentence-transformers
git checkout generic-logging
```

Finally, install the local version of `sentence-transformers` with `pip`
```
pip install -e .
```

Now once you import `sentence-transformers` you will be running with my modified version.

From there you should be ready to run! Let me know if you have any questions

**Comment by mpetruc** [comment]
@zachschillaci27 thank you for the patch for the installation instructions. Would it be too much to ask to give an example of how to implement it without wandb? I tried to modify the examples in #1606 to use logging instead of wandb, without much success. Also, how/where do you configure it to use the dev dataset for loss measurement? Thank you.

**Comment by mpetruc** [comment]
i'm still unable get any logging from the fit function. here's my code:
logging.basicConfig(filename="log/training.log",
    format='%(asctime)s - %(message)s',
                    datefmt='%Y-%m-%d %H:%M:%S',
                    level=logging.INFO)
loss_names = ["cosine_sim_loss"]
def log_callback_st(train_ix, epoch, training_steps, current_lr, loss_value):
        try:
            logging.info(
                    f"train/{loss_names[train_ix]}_epoch: {epoch} ",
                    f"train/{loss_names[train_ix]}_loss: {loss_value} ",
                    f"train/{loss_names[train_ix]}_lr: {current_lr} ",
                    f"train/steps: {training_steps}"
            )
        except Exception as e:
            print(f"Error in log_callback_st: {e}")
model = SentenceTransformer(modules=[word_embedding_model, pooling_model])
train_loss = losses.CosineSimilarityLoss(model=model)
model.fit(train_objectives=[(train_dataloader, train_loss)],
          evaluator=evaluator,
          epochs=num_epochs,
          evaluation_steps=1000, 
          warmup_steps=warmup_steps,
          output_path=model_save_path,
          use_amp=True,
          log_steps=len(train_dataloader)*num_epochs//1000, log_callback=log_callback_st)

The other logging.info statements work as expected.
What am i doing wrong?
thanks

**Comment by johnsonice** [comment]
any updates ??

File: sentence_transformers/SentenceTransformer.py [hunk 1]
```diff
@@ -728,6 +734,12 @@ def fit(self,
                     if not skip_scheduler:
                         scheduler.step()
 
+                    if log_steps > 0 and training_steps % log_steps == (log_steps - 1) and log_callback is not None:
+                        try:
+                            log_callback(train_idx, epoch, training_steps, scheduler.get_last_lr(), loss_value.item())
```
Comment (parth-jaggi-wd) [inline sentence_transformers/SentenceTransformer.py:739]: Can you also include `global_step` here?

**Comment by tomaarsen** [comment]
Hello!

Apologies for the delay in getting to this. The v3.0 release from last week has implemented a `SentenceTransformerTrainer` which accepts [`TrainerCallback` instances](https://sbert.net/docs/sentence_transformer/training_overview.html#callbacks) for extra callbacks for advanced logging. Additionally, the v3.0 released with better logging (e.g. integrations with Weights and Biases, Tensorboard, and better logging in the terminal).
In the future, this will be extended to CrossEncoders too.

As a result, I'll close this PR in favor of the v3.0 changes. I do want to thank you for your work on this.

- Tom Aarsen
